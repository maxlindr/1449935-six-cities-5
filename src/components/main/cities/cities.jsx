import React from 'react';
import PropTypes from 'prop-types';
import {offerPropTypes, cityPropTypes} from '../../../prop-types';
import OffersList from '../../offers-list/offers-list';
import SortDropdownList from './sort-dropdown-list/sort-dropdown-list';
import CityMap from '../../city-map/city-map';

const SortOption = {
  POPULAR: `popular`,
  TO_HIGH: `low-to-high`,
  TO_LOW: `high-to-low`,
  TOP_RATED: `top-rated`
};

const DROPDOWN_OPTIONS = {
  [SortOption.POPULAR]: `Popular`,
  [SortOption.TO_HIGH]: `Price: low to high`,
  [SortOption.TO_LOW]: `Price: high to low`,
  [SortOption.TOP_RATED]: `Top rated first`,
};

const onCardMouseOver = () => {};

const sortOffers = (offers, sortType) => {
  switch (sortType) {
    case SortOption.POPULAR:
      return offers;
    case SortOption.TO_HIGH:
      return offers.slice().sort((a, b) => a.price - b.price);
    case SortOption.TO_LOW:
      return offers.slice().sort((a, b) => b.price - a.price);
    case SortOption.TOP_RATED:
      return offers.slice().sort((a, b) => b.rating - a.rating);
    default:
      throw new Error(`Bad sort type: ${sortType}`);
  }
};

class Cities extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      sortType: SortOption.POPULAR,
    };

    this.onSortOptionChange = (option) => this.setState({sortType: option});
  }

  render() {
    const {city, offers} = this.props;
    const sortedOffers = sortOffers(offers, this.state.sortType);

    return (
      <div className="cities__places-container container">
        <section className="cities__places places">
          <h2 className="visually-hidden">Places</h2>
          <b className="places__found">
            {sortedOffers.length} places to stay in {city.name}
          </b>

          {/* При каждом рендере SortDropdownList должен закрываться, а для этого нужно сбрасывать его текущее состояние.
          Это решается с помощью генерации нового 'key' */}
          <SortDropdownList
            key={Math.random()}
            activeOption={this.state.sortType}
            options={DROPDOWN_OPTIONS}
            onChange={this.onSortOptionChange}
          />

          <OffersList
            offers={sortedOffers}
            onCardMouseOver={onCardMouseOver}
          />
        </section>

        <div className="cities__right-section">
          <CityMap city={city} offers={offers}/>
        </div>
      </div>
    );
  }

}

Cities.propTypes = {
  offers: PropTypes.arrayOf(offerPropTypes),
  city: cityPropTypes.isRequired
};

export default Cities;
