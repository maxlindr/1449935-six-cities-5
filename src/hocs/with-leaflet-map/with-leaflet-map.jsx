import React, {useCallback, useEffect, useRef, useState} from 'react';
import PropTypes from 'prop-types';
import {cityPropTypes, offerPropTypes} from '../../prop-types';
import leaflet from 'leaflet';
import 'leaflet/dist/leaflet.css';
import {connect} from 'react-redux';
import {getCurrentCity} from '../../store/selectors';

const PIN_SIZE = [27, 39];
const DEFAULT_ZOOM_LEVEL = 12;

const PIN = leaflet.icon({
  iconUrl: `img/pin.svg`,
  iconSize: PIN_SIZE
});

const PIN_ACTIVE = leaflet.icon({
  iconUrl: `img/pin-active.svg`,
  iconSize: PIN_SIZE
});

const checkOffersListsEquals = (offers1, offers2) => {
  if (offers1.length !== offers2.length) {
    return false;
  }

  return offers1.every((offer) => offers2.includes(offer));
};

const usePrevious = (value) => {
  const ref = useRef();

  useEffect(() => {
    ref.current = value;
  });

  return ref.current;
};

const withLeafletMap = (Component) => {
  const WithLeafletMap = (props) => {
    const {offers, city, activeOffer, maxOffers} = props;

    const mapRef = useRef(null);
    const [map, setMap] = useState(null);
    const [pinsGroup, setPinsGroup] = useState(null);
    const [mapPins] = useState(new Map());
    const [pinsData, setPinsData] = useState(new Map());
    const prevOffers = usePrevious(offers);
    const prevActiveOffer = usePrevious(activeOffer);

    useEffect(() => {
      return () => {
        mapPins.clear();
      };
    }, []);

    // reset map if city changed
    useEffect(() => {
      if (map) {
        map.setView(city.coordinates, DEFAULT_ZOOM_LEVEL);
        clear();
      }

    }, [city]);

    // update pins data
    useEffect(() => {
      if (
        activeOffer === prevActiveOffer &&
        checkOffersListsEquals(offers, prevOffers)
      ) {
        return;
      }

      const offersToShow = maxOffers
        ? offers.slice(0, maxOffers)
        : offers.slice();

      const newPinsData = new Map(
          offersToShow.map((offer) => {
            const {id, location} = offer;

            const data = {
              id,
              coordinates: location.coordinates,
              isActive: activeOffer ? (id === activeOffer.id) : false
            };

            return [id, data];
          })
      );

      setPinsData(newPinsData);
    }, [offers, activeOffer]);

    // draw pins
    useEffect(() => {
      if (!map) {
        return;
      }

      pinsData.forEach((pinData) => {
        const placedMapPin = mapPins.get(pinData.id);

        if (placedMapPin) {
          if (placedMapPin.isActive !== pinData.isActive) {
            updatePin(pinData);
          }
        } else {
          addPin(pinData);
        }
      });
    }, [map, pinsData]);

    // creating map
    useEffect(() => {
      const leafletMap = leaflet.map(mapRef.current, {
        center: city.coordinates,
        zoom: DEFAULT_ZOOM_LEVEL,
        zoomControl: false,
        marker: true
      });

      leaflet
        .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
          attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
        })
        .addTo(leafletMap);

      const leafletPinsGroup = leaflet.layerGroup();
      leafletPinsGroup.addTo(leafletMap);
      setPinsGroup(leafletPinsGroup);

      setMap(leafletMap);
    }, []);

    const addPin = useCallback(
        (pin) => {
          const {coordinates, isActive, id} = pin;

          const marker = leaflet.marker(coordinates, {
            icon: isActive ? PIN_ACTIVE : PIN
          });

          pinsGroup.addLayer(marker);

          mapPins.set(id, {
            pin,
            marker
          });
        },
        [pinsGroup, mapPins]
    );

    const clear = useCallback(
        () => {
          mapPins.clear();
          pinsData.clear();
          pinsGroup.clearLayers();
        },
        [mapPins, pinsData, pinsGroup]
    );

    const updatePin = useCallback(
        (pin) => {
          const oldPin = mapPins.get(pin.id);
          const icon = pin.isActive ? PIN_ACTIVE : PIN;
          oldPin.marker.setIcon(icon);
        },
        [mapPins]
    );

    return <Component reference={mapRef}/>;
  };

  WithLeafletMap.propTypes = {
    city: cityPropTypes.isRequired,
    offers: PropTypes.arrayOf(offerPropTypes).isRequired,
    activeOffer: offerPropTypes,
    maxOffers: PropTypes.number
  };

  return WithLeafletMap;
};

const mapStateToProps = (state) => ({
  city: getCurrentCity(state),
});

export {withLeafletMap};

export default (WrappedComponent) => connect(mapStateToProps)(
    withLeafletMap(WrappedComponent)
);
