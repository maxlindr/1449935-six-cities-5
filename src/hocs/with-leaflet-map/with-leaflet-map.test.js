import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import {withLeafletMap} from './with-leaflet-map';
import offers from '../../../__mocks__/offer-mocks';
import cities from '../../../__mocks__/city-mocks';

jest.mock(`leaflet`, () => {
  const doNothing = () => {};

  const layer = {
    addTo: doNothing,
    addLayer: doNothing,
  };

  return {
    icon: doNothing,
    map: doNothing,
    tileLayer: () => layer,
    layerGroup: () => layer,
    marker: doNothing,
  };
});

const MockComponent = () => <div>Mock Component</div>;

MockComponent.propTypes = {
  reference: PropTypes.shape({
    current: PropTypes.instanceOf(Element)
  }).isRequired,
};

const MockComponentWrapped = withLeafletMap(MockComponent);

describe(`withLeafletMap`, () => {
  it(`Map should render map component with pins`, () => {
    const tree = renderer
      .create(<MockComponentWrapped city={cities[0]} offers={offers} />, {
        createNodeMock: () => document.createElement(`div`),
      })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
