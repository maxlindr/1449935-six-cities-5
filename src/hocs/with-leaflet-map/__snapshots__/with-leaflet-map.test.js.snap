// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLeafletMap Map should render map component with pins 1`] = `
<div>
  Mock Component
</div>
`;
